src/Interpreters/QueryMetricLog.cpp

```diff
@@ -0,0 +1,180 @@
+#include <base/getFQDNOrHostName.h>
+#include <Common/CurrentThread.h>
+#include <Common/DateLUT.h>
+#include <Common/DateLUTImpl.h>
+#include <DataTypes/DataTypeDate.h>
+#include <DataTypes/DataTypeDateTime.h>
+#include <DataTypes/DataTypeDateTime64.h>
+#include <DataTypes/DataTypeLowCardinality.h>
+#include <DataTypes/DataTypeString.h>
+#include <DataTypes/DataTypesNumber.h>
+#include <Interpreters/Context.h>
+#include <Interpreters/QueryMetricLog.h>
+#include <Interpreters/PeriodicLog.h>
+#include <Interpreters/ProcessList.h>
+#include <Parsers/ExpressionElementParsers.h>
+#include <Parsers/parseQuery.h>
+
+#include <chrono>
+#include <mutex>
+
+namespace CurrentMetrics
+{
+    extern const Metric MemoryTracking;
+    extern const Metric MergesMutationsMemoryTracking;
+}
+
+namespace ErrorCodes
+{
+    extern const int LOGICAL_ERROR;
+}
+
+namespace DB
+{
+
+const auto memory_metrics = std::array{CurrentMetrics::MemoryTracking, CurrentMetrics::MergesMutationsMemoryTracking};
+
+ColumnsDescription QueryMetricLogElement::getColumnsDescription()
+{
+    ColumnsDescription result;
+    ParserCodec codec_parser;
+
+    result.add({"hostname",
+                std::make_shared<DataTypeLowCardinality>(std::make_shared<DataTypeString>()),
+                parseQuery(codec_parser, "(ZSTD(1))", 0, DBMS_DEFAULT_MAX_PARSER_DEPTH, DBMS_DEFAULT_MAX_PARSER_BACKTRACKS),
+                "Hostname of the server executing the query."});
+    result.add({"event_date",
+                std::make_shared<DataTypeDate>(),
+                parseQuery(codec_parser, "(Delta(2), ZSTD(1))", 0, DBMS_DEFAULT_MAX_PARSER_DEPTH, DBMS_DEFAULT_MAX_PARSER_BACKTRACKS),
+                "Event date."});
+    result.add({"event_time",
+                std::make_shared<DataTypeDateTime>(),
+                parseQuery(codec_parser, "(Delta(4), ZSTD(1))", 0, DBMS_DEFAULT_MAX_PARSER_DEPTH, DBMS_DEFAULT_MAX_PARSER_BACKTRACKS),
+                "Event time."});
+    result.add({"event_time_microseconds",
+                std::make_shared<DataTypeDateTime64>(6),
+                "Event time with microseconds resolution."});
+    result.add({"query_id",
+                std::make_shared<DataTypeString>(),
+                parseQuery(codec_parser, "(ZSTD(1))", 0, DBMS_DEFAULT_MAX_PARSER_DEPTH, DBMS_DEFAULT_MAX_PARSER_BACKTRACKS),
+                "Query ID."});
+
+    for (const auto & metric : memory_metrics)
+    {
+        auto name = fmt::format("CurrentMetric_{}", CurrentMetrics::getName(metric));
```

Comment:
If we are only going to save 2 metrics related to memory, maybe we can simply forget about CurrentMetrics and save the values in the element itself. WDYT?